facilitator/src/task/pubsub.rs

```diff
@@ -0,0 +1,185 @@
+                "https://www.googleapis.com/auth/pubsub",
+                identity.map(|x| x.to_string()),
+                None, // GCP key file; never used
+            )?,
+            phantom_task: PhantomData,
+        })
+    }
+}
+
+impl<T: Task> TaskQueue<T> for PubSubTaskQueue<T> {
+    fn dequeue(&mut self) -> Result<Option<TaskHandle<T>>> {
+        info!(
+            "pull task from {}/{} as {:?}",
+            self.gcp_project_id, self.subscription_id, self.oauth_token_provider
+        );
+
+        // API reference: https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull
+        let url = format!(
+            "{}/v1/projects/{}/subscriptions/{}:pull",
+            self.pubsub_api_endpoint, self.gcp_project_id, self.subscription_id
+        );
+
+        let http_response = ureq::post(&url)
+            .set(
+                "Authorization",
+                &format!("Bearer {}", self.oauth_token_provider.ensure_oauth_token()?),
+            )
+            // By default, ureq will wait forever to connect or read.
+            // Empirically, if there are no messages available in the
+            // subscription, the PubSub API will wait about 90 seconds to send
+            // an HTTP 200 with an empty JSON body. We set higher timeouts than
+            // usual to allow for this.
+            .timeout_connect(180_000) // three minutes
+            .timeout_read(180_000) // three minutes
+            .set("Content-Type", "application/json")
+            .send_json(ureq::json!({
+                // Dequeue one task at a time
+                "maxMessages": 1
+            }));
+        if http_response.error() {
+            return Err(anyhow!(
+                "failed to pull messages from PubSub topic: {:?}",
+                http_response
+            ));
+        }
+
+        let response = http_response
+            .into_json_deserialize::<PullResponse>()
+            .context("failed to deserialize response from PubSub API")?;
+
+        let received_messages = match response.received_messages {
+            Some(ref messages) => messages,
+            None => return Ok(None),
+        };
+
+        if received_messages.len() > 1 {
+            return Err(anyhow!(
+                "unexpected number of messages in PubSub API response: {:?}",
+                response
+            ));
+        }
+
+        if received_messages.len() == 0 {
+            return Ok(None);
+        }
+
+        // The JSON task is encoded as Base64 in the pubsub message
+        let task_json = base64::decode(&received_messages[0].message.data)
+            .context("failed to decode PubSub message")?;
+
+        let handle = TaskHandle {
+            task: T::from_json(&task_json)?,
+            acknowledgment_id: received_messages[0].ack_id.clone(),
+        };
+
+        Ok(Some(handle))
+    }
+
+    fn acknowledge_task(&mut self, handle: TaskHandle<T>) -> Result<()> {
```

Comment:
fn for nack or send to dead_letter queue is missing. I recommend a dead-letter queue, in that case you might want to differentiate between non-recoverable errors and recoverable errors. For non-recoverable errors you send to the dead-letter queue (which is a way to help investigate those issues without OOM (capacity-wise) the system with unprocessable tasks. For recoverable errors it would be good to keep a retry counter (when limit reached they are sent to a dead-letter queue as well) -- multiple ways you could implement that, either you ack the message and queue it with an updated retry counter, or you use your maker files to keep track of the counter.

Reply:
Done; I added the `nackowledge_task` method to the `TaskQueue` trait and implemented it here and in `AwsSqsTaskQueue`.